scripts/apps/monitoring/controllers/AggregateCtrl.js

```diff
@@ -87,25 +90,56 @@ export function AggregateCtrl($scope, api, desks, workspaces, preferencesService
+            .then(function(activeWorkspace) {
+                if (self.settings != null && self.settings.desk) {
+                    return deskMonitoringConfig(self.settings.desk);
+                } else if (activeWorkspace.type === 'workspace') {
+                    return preferencesService.get(PREFERENCES_KEY)
+                    .then(function(preference) {
+                        if (preference && preference[activeWorkspace.id] && preference[activeWorkspace.id].groups) {
+                            return {'type': 'workspace', 'groups': preference[activeWorkspace.id].groups};
+                        }
+                        return {'type': 'workspace', 'groups': []};
+                    });
+                } else if (activeWorkspace.type === 'desk') {
+                    // Read available groups from user preferences first.
+                    return preferencesService.get(PREFERENCES_KEY).then(function(preference) {
+                        let desk = self.deskLookup[activeWorkspace.id];
+                        let monitoringSettings = desk ? desk.monitoring_settings || [] : [];
+                        let activePrefGroups = preference[activeWorkspace.id] ? preference[activeWorkspace.id].groups || [] : [];
+
+                        if (activePrefGroups.length) {
+                            if (monitoringSettings.length) {
+                                // compare and determine if set of groups in desk monitoring settings & user preferences are same or changed
+                                // now, due to stages activated or deactivated in desk's monitoring settings.
+                                let diff = _.xorBy(monitoringSettings, activePrefGroups, '_id');
+
+                                if (diff.length) {
+                                    // if different, that means available stages/groups are changed now in desk monitoring settings
+                                    // so simply return recent desk monitoring settings.
+                                    return {'type': 'desk', 'groups': monitoringSettings};
+                                } else {
+                                    // update groups in preferences with any changes in desk's monitoring settings groups.
+                                    activePrefGroups.forEach(function(group) {
+                                        return angular.extend(group, monitoringSettings.find(grp => grp._id === group._id));
+                                    });
+
+                                    return {'type': 'desk', 'groups': activePrefGroups};
                                 }
-                                return {'type': 'workspace', 'groups': []};
-                            });
-                    } else if (activeWorkspace.type === 'desk') {
-                        var desk = self.deskLookup[activeWorkspace.id];
+                            }
+
+                            return {'type': 'desk', 'groups': activePrefGroups};
```

Comment:
This same `return` statement is repeated too many times. Can you just create a `groups` variable and assign whatever value is needed and instead only return once at the end? ie:
```js
return {'type': 'desk', 'groups': groups};
```